include/pstl_offload/internal/usm_memory_replacement_common.h

```diff
@@ -25,19 +26,104 @@
 namespace __pstl_offload
 {
 
+#if __linux__
+inline void*
+__original_aligned_alloc(std::size_t __alignment, std::size_t __size)
+{
+    using __aligned_alloc_func_type = void* (*)(std::size_t, std::size_t);
+
+    static __aligned_alloc_func_type __orig_aligned_alloc =
+        __aligned_alloc_func_type(dlsym(RTLD_NEXT, "aligned_alloc"));
+    return __orig_aligned_alloc(__alignment, __size);
+}
+#endif // __linux__
+
 constexpr bool
 __is_power_of_two(std::size_t __number)
 {
     return (__number != 0) && ((__number & __number - 1) == 0);
 }
 
+class __offload_policy_holder_type;
+
+// can't use std::shared_ptr, because commonly sizeof(std::shared_ptr) is 2*sizeof(void*) while
+// sizeof(__block_header) must be power of 2 and with std::shared_ptr instead of
+// __sycl_device_shared_ptr memory fragmentation increases drastically
+class __sycl_device_shared_ptr
+{
+    // state is shared between TU with the context and all objects allocated in this TU
+    struct __shared_device
+    {
+        sycl::device _M_device;
+        // to keep reference to default context of the device as long as allocated memory objects exist
+        sycl::context _M_default_context;
+        std::atomic<std::size_t> _M_cnt;
+    };
+
+    __shared_device* _M_shared_device;
+
+  public:
+    __sycl_device_shared_ptr() : _M_shared_device(nullptr) {}
+
+    explicit __sycl_device_shared_ptr(sycl::device _device)
+    {
+        _M_shared_device =
+            (__shared_device*)__original_aligned_alloc(alignof(std::max_align_t), sizeof(__shared_device));
+        if (!_M_shared_device)
+        {
+            throw std::bad_alloc();
+        }
+        new (_M_shared_device) __shared_device{_device, _device.get_platform().ext_oneapi_get_default_context(), 1};
+    }
+
+    bool
+    __is_device_created() const
+    {
+        return _M_shared_device != nullptr;
+    }
+
+    sycl::device
+    __get_device() const
+    {
+        return _M_shared_device->_M_device;
+    }
+
+    sycl::context
+    __get_context() const
+    {
+        return _M_shared_device->_M_default_context;
+    }
+
+    __sycl_device_shared_ptr&
+    operator=(const __sycl_device_shared_ptr& __other) = delete;
+
+    __sycl_device_shared_ptr(const __sycl_device_shared_ptr& __other)
```

Comment:
let's use member-init-list. That helps me not doing stupid mistakes by messing it with assignment :)

```suggestion
    __sycl_device_shared_ptr(const __sycl_device_shared_ptr& __other) : _M_shared_device(__other._M_shared_device)
```

Reply:
Done.